exo-run/src/app-runner.ls

```diff
@@ -56,17 +55,33 @@ class AppRunner extends EventEmitter
     DockerCompose.kill-all-containers {cwd: @docker-config-location, @write}, -> process.exit exit-code
 
 
-  _compile-online-text: ->
-    online-texts = {}
+  _compile-online-text: (done) ~>
+    @online-texts = {}
     for app-dependency in @app-config.dependencies
       dependency = ApplicationDependency.build app-dependency
-      online-texts[app-dependency.type] = dependency.get-online-text!
+      @online-texts[app-dependency.type] = dependency.get-online-text!
+    services = []
     for protection-level of @app-config.services
       for role, service-data of @app-config.services[protection-level]
-        if service-data.location #TODO: compile online text for external services
-          service-config = yaml.safe-load fs.read-file-sync(path.join(process.cwd!, service-data.location, 'service.yml'))
-          online-texts[role] = service-config.startup['online-text']
-    online-texts
+        services.push {role: role, service-data: service-data}
+    async.map-series services, (@_get-online-text), (err) ~>
+      | err => done err
+      done @online-texts
+
+
+  _get-online-text: ({role, service-data}, done) ~>
+    if service-data.location
+      service-config = yaml.safe-load fs.read-file-sync(path.join(process.cwd!, service-data.location, 'service.yml'))
+      @online-texts[role] = service-config.startup['online-text']
+      done!
+    else
+      switch
+        |!service-data['docker-image'] => throw new Error red "No location or docker-image specified"
+        DockerHelper.cat-file image: service-data['docker-image'], file-name: 'service.yml', (err, external-service-config) ~>
+          | err => throw new Error red "Could not find the configuration for the docker-image"
```

Comment:
`done err`, let's always bubble up our errors to the top level, and then we can print the error message there.